src/geopackage/sqlite/iterator.cpp

```diff
@@ -0,0 +1,152 @@
+#include <algorithm>
+
+#include "NGen_SQLite.hpp"
+
+using namespace geopackage;
+
+// Defined in database.cpp
+extern std::runtime_error sqlite_error(const std::string& f, int code, const std::string& extra = "");
+
+/**
+ * Runtime error for iterations that haven't started
+ */
+const auto sqlite_get_notstarted_error = std::runtime_error(
+    "sqlite iteration is has not started, get() is not callable (call sqlite_iter::next() before)"
+);
+
+/**
+ * Runtime error for iterations that are finished
+ */
+const auto sqlite_get_done_error = std::runtime_error(
+    "sqlite iteration is done, get() is not callable"
+);
+
+sqlite_iter::sqlite_iter(stmt_t stmt)
+  : stmt(stmt)
+{
+    // sqlite3_column_type requires the last result code to be
+    // SQLITE_ROW, so we need to iterate on the first row.
+    // TODO: need to test how this functions if the last result code
+    //       was SQLITE_DONE.
+    this->next();
+    this->column_count = sqlite3_column_count(this->ptr());
+    this->column_names = std::vector<std::string>();
```

Comment:
Default constructors have already run for any data member not named in the initializer list. There's no need to repeat initialization (and it may be a pessimization)